serialization_v2/serialization.py

```diff
@@ -0,0 +1,1243 @@
+            sis_path_handling=SisPathHandling.AS_STRING,
+        )
+        return "\n\n".join(
+            part
+            for part in [
+                "#!returnn/rnn.py",
+                python_prolog_code,
+                serialized.as_serialized_code(),
+                python_epilog_code,
+                "# -*- mode: python; tab-width: 4 -*-\n",
+            ]
+            if part.strip()
+        )
+
+
+@dataclass
+class SerializedConfig:
+    code_list: List[PyCode]
+
+    def as_serialization_collection(self) -> SerializerCollection:
+        """as serialization Collection"""
+        return SerializerCollection(self.code_list)
+
+    def as_serialized_code(self) -> str:
+        """as serialized code"""
+        return "".join(code.py_code for code in self.code_list)
+
+
+class _Serializer:
+    def __init__(
+        self,
+        config: Dict[str, Any],
+        post_config: Optional[Dict[str, Any]] = None,
+        known_modules: Collection[str] = (),
+        sis_path_handling: SisPathHandling = None,
+    ):
+        # from returnn.tensor import Dim, batch_dim, single_step_dim
+
+        self.config = config.copy()
+        self.post_config = post_config.copy() if post_config else {}
+        self.assignments_dict_by_value_ref: Dict[_Ref, PyCode] = {}  # value ref -> code
+        self.assignments_dict_by_name: Dict[str, PyCode] = {}  # var name -> code
+        self.assignments_dict_by_idx: Dict[int, PyCode] = {}  # idx -> code
+        self.assignments_dict_by_value_by_type: Dict[type, Dict[Any, PyCode]] = {
+            # TODO: is this entry needed for correctness?
+            # Dim: {},
+        }  # type -> dict value -> code
+        self.reduce_cache_by_value_ref: Dict[_Ref, Tuple[Any, ...]] = {}  # value ref -> (func, args, ...)
+        self.added_sys_paths = set()
+        self.known_modules = set(known_modules)
+        for mod_name in known_modules:
+            mod = sys.modules[mod_name]
+            # Don't add those module path to sys.path again.
+            self.added_sys_paths.add(_get_module_path_from_module(mod))
+        self._next_sys_path_insert_idx = 0
+        self.sis_path_handling = sis_path_handling
+        self._cur_added_refs: List[PyCode] = []
+        self._next_assignment_idx = 0
+        # We first serialize everything without inlining anything.
+        # There we also count how often a value is used (ref_count).
+        # Then we can inline those values which are not direct config entries
+        # and which are only used once.
+        self._inlining_stage = False
+
+        # Avoid to use them, but if necessary (when inside the config), they can be used.
+        # The config keys always have precedence.
+        self._internal_reserved_names = {
+            "sys": sys,
+            # TODO: are these entries needed for correctness?
+            # "batch_dim": batch_dim,
+            # "single_step_dim": single_step_dim,
+            # "Dim": Dim,
```

Comment:
I think these are not needed. I would however maybe adapt the code for suggesting the variable name to try to keep the original name here.